jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java

```diff
@@ -57,15 +60,26 @@ public HostHeaderCustomizer(String serverName)
      */
     public HostHeaderCustomizer(String serverName, int serverPort)
     {
-        this.serverName = Objects.requireNonNull(serverName);
+        this.serverName = serverName;
         this.serverPort = serverPort;
     }
 
     @Override
     public void customize(Connector connector, HttpConfiguration channelConfig, Request request)
     {
-        if (request.getHeader("Host") == null)
-            // TODO set the field as well?
-            request.setHttpURI(HttpURI.build(request.getHttpURI()).host(serverName).port(serverPort));
+        if (request.getHttpVersion() != HttpVersion.HTTP_1_1 && !request.getHttpFields().contains(HttpHeader.HOST))
+        {
+            String host = serverName == null ? request.getServerName() : serverName;
+            int port = HttpScheme.normalizePort(request.getScheme(), serverPort == 0 ? request.getServerPort() : serverPort);
+
+            if (serverName != null || serverPort > 0)
+                request.setHttpURI(HttpURI.build(request.getHttpURI()).authority(host, port));
+
+            HttpFields original = request.getHttpFields();
+            HttpFields.Mutable httpFields = HttpFields.build(original.size() + 1);
+            httpFields.add(new HostPortHttpField(host, port));
+            httpFields.add(request.getHttpFields());
```

Comment:
This looks odd, why is this not `httpFields.addAll(request.getHttpFields());` ?

Also, doesn't this change the `request.getHttpFields()` to now be Mutable?

What if this was ...

``` java
HttpFields updated = HttpFields.build(
   request.getHttpFields(), 
   new HostPortHttpFields(host, port));
request.setHttpFields(updated);
```

Where there is a new method `.build(HttpFields original, HttpField ... fields)` ?

Reply:
https://tools.ietf.org/html/rfc7230#section-3.2.2 says Host header should be one of the first headers sent.  I can't imagine too much depending on this, but best to follow recommendations.

Reply:
This still changes future requests on `request.getHttpFields()` to be a `HttpFields.Mutable`, right?
I don't think that's desired.

Reply:
https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java#L250-L253